client/src/telethon/_impl/mtsender/sender.py

```diff
@@ -239,56 +237,52 @@ async def _step_until_receive(self, rx: Future[bytes]) -> bytes:
                 return rx.result()
 
     async def step(self) -> None:
-        ticket_number = self._step_counter
+        if not self._writing:
+            self._writing = True
+            await self._do_write()
+            self._writing = False
 
-        async with self.lock:
-            if self._step_counter == ticket_number:
-                # We're the one to drive IO.
-                self._step_counter += 1
-                await self._step()
-            # else: different task drove IO.
+        if not self._reading:
+            self._reading = True
+            await self._do_read()
+            self._reading = False
+
+        if not self._step_done.is_set():
+            await self._step_done.wait()
 
     def pop_updates(self) -> list[Updates]:
         updates = self._updates[:]
         self._updates.clear()
         return updates
 
-    async def _step(self) -> None:
-        self._try_fill_write()
-
-        recv_req = asyncio.create_task(self._request_event.wait())
-        recv_data = asyncio.create_task(self._reader.read(MAXIMUM_DATA))
-        send_data = asyncio.create_task(self._do_send())
-        done, pending = await asyncio.wait(
-            (recv_req, recv_data, send_data),
-            timeout=self._next_ping - asyncio.get_running_loop().time(),
-            return_when=FIRST_COMPLETED,
-        )
+    async def _do_read(self) -> None:
+        self._step_done.clear()
 
-        if pending:
-            for task in pending:
-                task.cancel()
-            await asyncio.wait(pending)
-
-        if recv_req in done:
-            self._request_event.clear()
-        if recv_data in done:
-            self._on_net_read(recv_data.result())
-        if send_data in done:
-            self._on_net_write()
-        if not done:
-            self._on_ping_timeout()
-
-    async def _do_send(self) -> None:
-        if self._write_drain_pending:
-            await self._writer.drain()
-            self._write_drain_pending = False
+        try:
+            async with asyncio.timeout(PING_DELAY):
+                recv_data = await self._reader.read(MAXIMUM_DATA)
+        except TimeoutError:
+            pass
         else:
-            # Never return
-            await asyncio.get_running_loop().create_future()
+            self._on_net_read(recv_data)
+        finally:
+            self._try_timeout_ping()
+            self._step_done.set()
 
-    def _try_fill_write(self) -> None:
-        if self._write_drain_pending:
+    async def _do_write(self) -> None:
+        self._step_done.clear()
+
+        try:
+            async with asyncio.timeout(PING_DELAY):
```

Comment:
Can both write and read timeout at the same time?

Perhaps it's better to only leave the ping timeout on either send or read, but not both. I would prefer read, because that's where we will block the most.